add location tabs to heatmap bottom view port

yuncmsft opened the pull request:
need to add three location tabs to heatmap bottom view port (currently one table showing all source IPs/query counts). * discussion and feedback needed *

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

yuncmsft committed:
add location tabs to heatmap bottom view port

need to add three location tabs to heatmap bottom view port (currently one table showing all source IPs/query counts).

azuresdkci commented:
Can one of the admins verify this patch?

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-java

AutorestCI commented:
# Automation for azure-sdk-for-go

yuncmsft committed:
add location tabs to heatmap bottom view port

[change type to number] 
Need to add three location tabs under current Traffic Manager Heatmap view port.  [discussion and feedback needed]

yuncmsft committed:
add location tabs to heatmap bottom view port

in addition to default heatmap table, add three location options

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=3559) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=3559) before we can accept your contribution.<br/><br/>:x: yuncmsft [sign now](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=3559)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/azure-rest-api-specs?pullRequest=3559) it.</sub>

lmazuel commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -729,6 +729,9 @@
             "type": "string",
             "enum": [
               "default"
```
Missing comma

yuncmsft committed:
add location tabs to heatmap bottom view port

missing comma for heatMapType enum.

yuncmsft committed:
add location tabs to heatmap bottom view port 

add HeatMapPropertyBase and four sub-classes: default, asn, country, state;
add links to JSON examples
update HeatMapModel

yuncmsft committed:
HeatMapCountry-GET.json

example of HeatMapCountry GET (heapMapType == "country")

yuncmsft committed:
create HeatMapAsn-GET.json

example for HeatMapASN GET responses (heatMapType =="asn")

yuncmsft committed:
create HeatMapState-GET.json

an example for HeatMapState GET response (heatMapType == "state")

lmazuel commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -1098,34 +1107,111 @@
+        "heatMapType"
+      ],
+    },
+    "default":{
```
Use a meaingful name, and the `x-ms-discriminator-value` for the real value. Example:
```json
    "DatasetBZip2Compression": {
      "x-ms-discriminator-value": "BZip2",
```
This means "Create class DatasetBZip2Compression, but match value BZip2 on the wire"

yuncmsft replied:
I see. so here I can use  
"HeatMapPropertyAsn": 
   {"x-ms-discriminator-value": "asn",
 to indicate that when heatMapType=="asn", the HeatMapPropertyAsn will be applied? 
if so, that's very convenient - I was very confused how I can link the two.

lmazuel commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -1098,34 +1107,111 @@
+        }
+      ]
+    },
+    "asn":{
```
Use a meaingful name, and the `x-ms-discriminator-value` for the real value. Example:
```json
    "DatasetBZip2Compression": {
      "x-ms-discriminator-value": "BZip2",
```
This means "Create class DatasetBZip2Compression, but match value BZip2 on the wire"

lmazuel commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -1098,34 +1107,111 @@
+        }
+      ]
+    },
+    "country":{
```
Use a meaingful name, and the `x-ms-discriminator-value` for the real value. Example:
```json
    "DatasetBZip2Compression": {
      "x-ms-discriminator-value": "BZip2",
```
This means "Create class DatasetBZip2Compression, but match value BZip2 on the wire"

lmazuel commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -1095,34 +1107,111 @@
+      },
+      "required": [
+        "heatMapType"
+      ],
```
Extra comma, invalid JSON

lmazuel commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -1095,34 +1107,111 @@
-      "description": "Class representing a Traffic Manager HeatMap properties."
+      ]        
+    },
+    "state":{
```
Use a meaningful name, and the `x-ms-discriminator-value` for the real value. Example:
```json
    "DatasetBZip2Compression": {
      "x-ms-discriminator-value": "BZip2",
```
This means "Create class DatasetBZip2Compression, but match value BZip2 on the wire"

lmazuel commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -1137,6 +1226,19 @@
       },
       "description": "Class which is a sparse representation of a Traffic Manager endpoint."
     },
+    "HeatMapLocation": {
```
Now you split everything, this type should be used somewhere.

yuncmsft replied:
ar... forgot to update this def name, should be "TrafficLocation" - i'm assuming when I use "$ref": "#/definitions/TrafficLocation", this "TrafficLocation" will be applied.

yuncmsft committed:
update x-ms-discriminator-value

"default" -> HeatMapProperties
"asn" -> HeatMapPropertyAsn
"country" -> HeatMapPropertyCountry
"state" -> HeatMapPropertyState
rename HeatMapLocation to TrafficLocation (referenced)

yuncmsft committed:
Update HeatMapAsn-GET.json

update resource type name

yuncmsft committed:
Update HeatMapCountry-GET.json

update resource type

yuncmsft committed:
Update HeatMapState.json

update resource type

yuncmsft committed:
update x-ms-discriminator-value to type

update discriminator value to be resource type

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/examples/HeatMapState-GET.json:
```diff
@@ -0,0 +1,38 @@
+  "responses": {
+    "200": {
+      "body": {
+        "id": "/subscriptions/{subscription-id}/resourceGroups/azuresdkfornetautoresttrafficmanager1323/providers/Microsoft.Network/trafficManagerProfiles/azuresdkfornetautoresttrafficmanager3880/heatMaps/stateTrafficFraction",
```
name value coming from "id" does not match the value in "name" property. Please make sure service returns the correct "id" and "name" value

yuncmsft replied:
updated name and type - do we need a separate API version?

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/examples/HeatMapCountry-GET.json:
```diff
@@ -0,0 +1,38 @@
+      "body": {
+        "id": "/subscriptions/{subscription-id}/resourceGroups/azuresdkfornetautoresttrafficmanager1323/providers/Microsoft.Network/trafficManagerProfiles/azuresdkfornetautoresttrafficmanager3880/heatMaps/countryTrafficFraction",
+        "name": "country",
+        "type": "Microsoft.Network/trafficManagerProfiles/heatMaps/countryTrafficFraction",
```
name, id, type values are all messed up. Please set up a sync with armapireview to get this approved.

ravbhatnagar reviewed (commented):
some noticeable issues. Lets close over a quick call. I want to make sure service is doing the right thing here.

ravbhatnagar commented:
LOoks like you have decided to go with enum for heatmap type with fixed set of names that can be given instead of $expand. Looks fine to me.

lmazuel commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -1095,34 +1107,115 @@
-    "HeatMapProperties": {
+    "HeatMapPropertyBase": {
+      "type": "object",
+      "discriminator": "heatMapType",
```
That's not exactly the way to do it, but we're close :)
- HeatMapModel should declare id, name, and type and nothing more
- HeatMapModel should have the discriminator on "type", since this is the actual value that discriminates (`heatMapType` does not exist in the JSON).
- For each type, you need two classes (take Country as example)
  - HeatMapModelCountry that is a allOf of HeatMapModel and has the x-discriminator-value, and contains one node properties described in a class HeatMapModelCountryProperties. This last one does not talk about discriminator

Hope this clarify a bit, it's easier to see now we made progress and have examples.

yuncmsft committed:
update example id/name/type

yuncmsft committed:
update example id/name/type

yuncmsft committed:
update example id/name/type

yuncmsft committed:
update discriminator to use 'name'

- Include ARM's comments to update name, and type. 
- Swagger discriminator now uses 'name' as distinguisher:  "default", "asnTrafficFraction", "countryTrafficFraction", "stateTrafficFraction".
- Updated existing HeatMapModel to be the super class with four dervatives with a discriminator value in each class: HeatMapModelDefault, HeatMaptModelAsn, HeatMapModelCountry, HeatMapModelState. 
- Each sub-model has a corresponding properties class

yuncmsft committed:
remove extra comma

remove extra comma

lmazuel committed:
Add generate-empty-classes

salameer commented:
@lmazuel  do you recall what's the deal with this one? should it be merged?

salameer commented:
ping @lmazuel  :)

dsgouda commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -1137,6 +1276,19 @@
       "description": "Class which is a sparse representation of a Traffic Manager endpoint."
     },
+    "TrafficLocation": {
+      "properties": {
```
Please mark the readonly properties wherever applicable

dsgouda reviewed (changes requested):
Looks good for the most part

salameer commented:
@yuncmsft can you please address the changes requested?

salameer commented:
CLosing since there's no response from @yuncmsft please request to reopen if required

AutorestCI commented:
# Automation for azure-sdk-for-js
